src/NetSapiensSharp/Connector.cs

```diff
@@ -26,51 +27,47 @@ public Connector(string api_base_url, string client_id, string client_secret, st
             _Password = password;
         }
 
-        private void AuthenticateUsernamePassword()
-        {
-            var r = Connector.Authenticate(_ApiBaseUrl, _ClientId, _ClientSecret, _Username, _Password);
-            if (r.Data == null)
-            {
-                throw (r.ErrorException);
-            }
-            _SessionToken = r.Data.access_token;
-            _RefreshToken = r.Data.refresh_token;
-            _ExpiresInSeconds = r.Data.expires_in;
-            _ExpirationStartTime = DateTime.UtcNow;
-        }
-
-        private void AuthenticateRefreshToken()
-        {
-            var r = Connector.Authenticate(_ApiBaseUrl, _ClientId, _ClientSecret, _RefreshToken);
-            if (r.Data == null)
-            {
-                throw (r.ErrorException);
-            }
-            _SessionToken = r.Data.access_token;
-            _RefreshToken = r.Data.refresh_token;
-            _ExpiresInSeconds = r.Data.expires_in;
-            _ExpirationStartTime = DateTime.UtcNow;
-        }
-
         private void Authenticate()
         {
             if (_SessionToken != null)
             {
                 return;
             }
-            if (_RefreshToken != null)
+
+            IRestResponse<Authentication_Response> myResponse = null;
+            int myRetryCount = 0;
+
+            do
             {
-                try
+                if (myRetryCount > 0)
```

Comment:
I know why we started down decomposing this from three methods into one larger method, but it seems quite a bit different in the details of how it is doing this now (aside from just retry).

Did we try adapting the old to use retries first? That might make it easier to demonstrate that only pieces surrounding retry is changing.

Reply:
I originally had designed both `AuthenticateUsernamePassword` and `AuthenticateRefreshToken` to each have their own retry logic but it added a bunch of duplicate code which is why I went with the approach of trying to combine the authentication functions.

Reply:
Understood - heading down a path of one larger function is generally a regression, so if we change anything here at all (not the direction I am recommending, mind you), then we probably want to think through how to do this in a way that is more clear in how it's approaching the solution.